[ZEPPELIN-1619] Load js package as a plugin visualization

Leemoonsoo opened the pull request:
### What is this PR for?
Current helium plugin application api (experimental) requires create library in java class, and need to create both backend / frontend code in the package. Which is good if your plugin requires both frontend and backend code running.

However, when user just want to make new visualization which completely runs on front-end side in javascript, creating helium application in java project and taking care of backend code can be bit of overhead and barrier for javascript developers.

This PR adds capability to load pure javascript package as a visualization.

### how it works

1. create (copy, download) 'helium package json' file into `ZEPPELIN_HOME/helium` directory.
  The json file point visualization js package in npm repository or local file system in `artifact` field.   
  `type` field in the json file need to be `VISUALIZATION`

Here's an example (zeppelin-examples/zeppelin-example-horizontalbar/zeppelin-example-horizontalbar.json)
```
{
  "type" : "VISUALIZATION",
  "name" : "zeppelin_horizontalbar",
  "description" : "Horizontal Bar chart (example)",
  "artifact" : "./zeppelin-examples/zeppelin-example-horizontalbar",
  "icon" : "<i class='fa fa-bar-chart rotate90flipX'></i>"
}
```

2. Go to helium GUI menu. (e.g. http://localhost:8080/#/helium).
  The menu will list all available packages. 
<img width="796" alt="writing_visualization_helium_menu" src="https://cloud.githubusercontent.com/assets/1540981/21749660/0f401c10-d558-11e6-9961-b6d0a9c023d8.png">

 
3. click 'enable' in any package want to use. 
Once a visualization package is enabled, `HeliumVisualizationFactory` will collect all enabled visualizations and create js bundle on the fly.

4. js bundle will be loaded on notebook and additional visualization becomes available
![image](https://cloud.githubusercontent.com/assets/1540981/21749729/709b2b3e-d559-11e6-8318-7f2871e7c39a.png)


### Programming API to create new plugin visualization.

Simply extends [visualization.js](https://github.com/apache/zeppelin/blob/master/zeppelin-web/src/app/visualization/visualization.js) and overrides some methods, such as

```
  /**
   * get transformation
   */
  getTransformation() {
    // override this
  };

  /**
   * Method will be invoked when data or configuration changed
   */
  render(tableData) {
    // override this
  };

  /**
   * Refresh visualization.
   */
  refresh() {
    // override this
  };

  /**
   * method will be invoked when visualization need to be destroyed.
   * Don't need to destroy this.targetEl.
   */
  destroy() {
    // override this
  };

  /**
   * return {
   *   template : angular template string or url (url should end with .html),
   *   scope : an object to bind to template scope
   * }
   */
  getSetting() {
    // override this
  };
```

This is exactly the same api that built-in visualization uses.

an example implementation included `zeppelin-examples/zeppelin-example-horizontalbar/horizontalbar.js`.
Actually [all built-in visualizations](https://github.com/apache/zeppelin/tree/master/zeppelin-web/src/app/visualization/builtins) are example


### Packaging and publishing visualization

Each visualization will need `package.json` file (e.g. `zeppelin-examples/zeppelin-example-horizontalbar/package.json`) to be packaged.
Package can be published in npm repository or package can be deployed to the local filesystem.

`zeppelin-examples/zeppelin-example-horizontalbar/` is an example package that is deployed in the local filesystem

### Development mode

First, locally install and enable your development package by setting `artifact` field to the development directory.
Then run zeppelin-web in visualization development mode with following command
```
cd zeppelin-web
npm run visdev
```
When you have change in your local development package, just reload your notebook. Then Zeppelin will automatically rebuild / reload the package.



Any feedback would be appreciated!


### What type of PR is it?
Feature

### Todos
* [x] - Load plugin visualization js package on runtime
* [x] - Make the feature works in zeppelin Binary package
* [x] - Show loading indicator while 'enable' / 'disable' package
* [x] - Add document
* [x] - Add license of new dependency
* [x] - Development mode
* [x] - Propagate error to front-end
* [x] - Display multiple versions of a package.

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-1619

### How should this be tested?
Build Zeppelin with `-Pexamples` flag. That'll install example visualization package `horizontalbar`.
You'll able to select `horizontalbar` along with other built-in visualizations
![image](https://cloud.githubusercontent.com/assets/1540981/21655057/27d61740-d26d-11e6-88f2-02c653e102c6.png)

To test npm online package install capability,  Place [zeppelin-bubble.json](https://github.com/Leemoonsoo/zeppelin-bubble/blob/master/zeppelin-bubble.json) in hour local registry (`ZEPPELIN_HOME/helium`) and enable it in Helium gui menu.
And then zeppelin will download package from npm repository and load.
![bubblechart](https://cloud.githubusercontent.com/assets/1540981/21749717/280aa430-d559-11e6-9209-889a4f86d7e2.gif)


### Questions:
* Does the licenses files need update? yes
* Is there breaking changes for older versions? no
* Does this needs documentation? yes

jongyoul commented:
@Leemoonsoo I like this idea to show new menu for helium but how about creating a new module like zeppelin-helium? It would be more helpful to understand whole structure. What do you think of it?

Leemoonsoo commented:
@jongyoul I think helium related code can be extracted from zeppelin-zengine to a new module, zeppelin-helium. However, extracting code and create new module is not this PR's scope. that can be handled in separate issue.

jongyoul commented:
It's ok of you to have a plan to divide them. Thanks.

Leemoonsoo commented:
Ci is green. I think this is ready. Please review this.

cloverhearts commented:
Awesome! LGTM!

1ambda commented:
Let me review this PR. I wil comment again :)

soralee commented:
It's very cool feature!
Tested and It works very nice like below.

**[Test result]**
When I clicked `Enable` button for using `zeppelin_horizontalbar` visualization package. 
![z1842_e](https://cloud.githubusercontent.com/assets/8110458/21835438/50aaa7ba-d800-11e6-9a4e-5dcbc6f4ffd4.gif)

When using `Disable` button 
![z1842_d](https://cloud.githubusercontent.com/assets/8110458/21835586/85b63f5e-d801-11e6-9784-f5dbc1fa1f21.gif)

soralee commented:
And when I test npm online package for using bubble chart, It works very well!
![z1842_b](https://cloud.githubusercontent.com/assets/8110458/21835960/94d69aee-d804-11e6-8b4f-8b51c0bf72cc.gif)

LGTM!

AhyoungRyu commented:
@Leemoonsoo Tested and it works well. It's really awesome! And 👍 for having really well described docs. Let me create my own viz application as explained in the documentation with real user's point of view. I will leave another comment after that :)

AhyoungRyu commented:
When I enable zero or only one vizualization, 
<img width="980" alt="screen shot 2017-01-12 at 12 13 32 am" src="https://cloud.githubusercontent.com/assets/10060731/21853851/fd09d836-d85b-11e6-99ee-4320e309676c.png">

I didn't no that there is reordering feature. It'll be better to show some messages in this case as well so that ppl can know about this. What do you think? 

And when I tried to reorder the icons, it doesn't work `left -> right`. It only works `right -> left` like below. 
![record](https://cloud.githubusercontent.com/assets/10060731/21854089/cffc4c6a-d85c-11e6-9f6a-74b03f74b02a.gif)

AhyoungRyu commented on docs/development/writingzeppelinvisualization.md:
```diff
@@ -0,0 +1,200 @@
+
+To create your own visualization, you need to create a js file and import [Visualization](https://github.com/apache/zeppelin/blob/master/zeppelin-web/src/app/visualization/visualization.js) class from [zeppelin-vis](https://github.com/apache/zeppelin/tree/master/zeppelin-web/src/app/visualization) package and extend the class. [zeppelin-tabledata](https://github.com/apache/zeppelin/tree/master/ze ...
+
+[Visualization](https://github.com/apache/zeppelin/blob/master/zeppelin-web/src/app/visualization/visualization.js) class, there're several methods that you need to override and implement. Here's simple visualition that just prints `Hello world`.
```
`visualition` -> `visualization `  :)

AhyoungRyu commented on docs/development/writingzeppelinvisualization.md:
```diff
@@ -0,0 +1,200 @@
+#### 2. Enable packages
+Once Zeppelin loads _Helium package files_ from local registry, available packages are displayed in Helium menu.
+
+User simply click 'enable' button to enable.
```
How about "Just click 'enable' button to enable them" ?

AhyoungRyu commented on docs/development/writingzeppelinvisualization.md:
```diff
@@ -0,0 +1,200 @@
+
+
+#### 1. Load Helium package files from registry
+Zeppelin need to know what Visualization packages are available. Zeppelin searches _Helium package file_ from local registry (by default helium/ directory) by default.
```
- Zeppelin `need` -> `needs`

AhyoungRyu commented on docs/development/writingzeppelinvisualization.md:
```diff
@@ -0,0 +1,200 @@
+
+##### artifact
+
+Localtion of the visualization npm package. Support npm package with version or local filesystem path.
```
`Localtion` -> `Location`

AhyoungRyu commented on docs/development/writingzeppelinvisualization.md:
````diff
@@ -0,0 +1,200 @@
+
+```
+cd zeppelin-web
+npm run visdev
````
How about `yarn run visdev` for the consistency with other commands listed in [zeppelin-web/README.md](https://github.com/apache/zeppelin/tree/master/zeppelin-web#local-development)?

AhyoungRyu commented on docs/rest-api/rest-helium.md:
````diff
@@ -0,0 +1,378 @@
+    <col width="200">
+    <tr>
+      <td>Description</td>
+      <td>This ```POST``` method set visualization packages display order.</td>
````
`set` -> `sets`

AhyoungRyu commented on zeppelin-web/src/app/helium/helium.controller.js:
```diff
@@ -0,0 +1,190 @@
+        closeByBackdrop: false,
+        closeByKeyboard: false,
+        title: '',
+        message: 'Enable ' + name + '? <div style="color:gray">' + artifact + '</div>',
```
Probably it's nit; how about "Do you want to enable + name + ?" for more kind message? :D

AhyoungRyu commented on zeppelin-web/src/app/helium/helium.controller.js:
```diff
@@ -0,0 +1,190 @@
+        closeByBackdrop: false,
+        closeByKeyboard: false,
+        title: '',
+        message: 'Disable ' + name + '?',
```
Same here :)

Leemoonsoo commented:
Thanks for the reviews and verifications!

@AhyoungRyu Addressed all the comments and fixed drag and drop problem.
When there is 0 or 1 visualization, reorder doesn't make any sense and i think we don't need to bother user with reorder feature in this case.

AhyoungRyu commented:
@Leemoonsoo Thanks! I just suggested to have some default message like other pages (e.g. localhost:8080/#/interpreter -> there is "Manage interpreters settings. You can create / edit / remove settings. Note can bind / unbind these interpreter settings.")But it's just minor suggestion :)

I tried to create new visualization package using HighCharts as you described in documentation, I didn't face any problem and it's definitely easier to create than before 👍 
![record](https://cloud.githubusercontent.com/assets/10060731/21900170/3a772144-d936-11e6-849d-ce047bb57ccc.gif)

Leemoonsoo commented:
@AhyoungRyu Thanks for the review. Great to see High chart working nicely.

A lot more improvements on Helium menu will be required, like search page, grouping packages by type, package list sort, etc. I'd like to address further improvements on Helium menu in separate PR.

Leemoonsoo commented:
Merge to master if no more comments!

jongyoul commented:
Good job! @Leemoonsoo
